### ISSUE TITLE
When an actor is passed as a part of a struct, it isn't an `Actor` instance

### ISSUE THREAD

**Issue by vporton** [dfinity/agent-js#775]
**Describe the bug**
When an actor is passed as a part of a struct, it isn't an `Actor` instance (but should be).

**To Reproduce**
Create a Motoko and a JavaScript file ([this repo](https://github.com/vporton/bug-actor-in-struct)):

Motoko:
```motoko
actor class canvalue_backend() = this {
  public type R = {
    value: canvalue_backend;
  };
  public query func test() : async R {
    return { value = this };
  };
};
```

TypeScript:
```typescript
import { Actor } from "@dfinity/agent";
import { canvalue_backend } from "../../declarations/canvalue_backend";

const ac = await canvalue_backend.test();
Actor.canisterIdOf(ac.value)
```
Error:
```text
Argument of type 'canvalue_backend' is not assignable to parameter of type 'Actor'.
  Property '[metadataSymbol]' is missing in type 'canvalue_backend' but required in type 'Actor'.
```

Generate:
```
dfx generate
```

Inspect the TypeScript code `src/canvalue_frontend/src/index.ts` in a text editor (such as MS Code) capable to spot TypeScript type errors.

**Expected behavior**
`ac` should be an instance of `Actor`.

**Desktop (please complete the following information):**

- OS: Ubuntu Linux
- Version 23.04

**Comment by vporton** [comment]
> It's not an Actor instance.

But it _should be_ an Actor instance. This is what I ask in my bug report.

**Comment by chenyan-dfinity** [comment]
You can construct an Actor from the principal, but it's current not done automatically. It can be a feature request, but it's not a bug, the current format is specified in the doc.

**Comment by vporton** [comment]
> You can construct an Actor from the principal

No, `ac` in my original message isn't a `Principal`. It is _some_ value that I have no idea how to use it.

Moreover,

```motoko
  public query func getRootItem(): async ?(CanDBPartition.CanDBPartition, Nat) {
    rootItem;
  };
```
(here `CanDBPartition` is an actor class).

```typescript
    main.getRootItem().then((data) => {
    })
```
crashed with the browser console message `Uncaught (in promise) Error: Illegal service definition: services can only contain functions`. So, it apparently crashes on any access to the value. It **is** a bug.

**Comment by vporton** [comment]
The reverse case of passing an actor as an argument does work, but its value needs to be converted into a principal before passing. This seems to work:

```motoko
actor class canvalue_backend() = this {
  public query func test(value: canvalue_backend) : async () {
  };
};
```

```typescript
import { Actor } from "@dfinity/agent";
import { canvalue_backend } from "../../declarations/canvalue_backend";

await canvalue_backend.test(Actor.canisterIdOf(canvalue_backend));
```

A bad thing here is that if we allow (as it would be in an ideal world) passing directly an actor, this would create past compatibility problems (that however can be solved by expecting `Actor | Principal`.

**Comment by chenyan-dfinity** [comment]
> No, ac in my original message isn't a Principal. It is some value that I have no idea how to use it.

`ac` is a record. `ac.value` is a principal. If not, can you `console.log(ac.value)`?
